packages/components/CONTRIBUTING.md

```diff
@@ -236,6 +232,53 @@ A couple of good examples of how hooks are used for composition are:
 
 TDB -->
 
+## Naming Conventions
```

Comment:
Maybe there was some nuance lost in the discussion, but my understanding was that things are not going to be in a clear "pure monolith vs. fully modular" binary. More like:

- _Relatively_ monolithic: This could be a pure monolith if it's a very simple component, but usually they are going to be at least somewhat modular. The `CustomSelectControlV2` API is a good example of this. It's modular, but isn't as low-level modular as the raw Ariakit API (`Ariakit.Select`, `Ariakit.SelectPopover`).
- Fully modular: I basically think of this as a mirror of the raw Ariakit API.

So I was thinking, the default API ("relatively monolithic") that most consumers will use is going to be:

```jsx
		<Component>
			<Component.SubComponent />
		</Component>
```

And the low-level API ("fully modular"), if the need arises to provide it, is going to look like:

```jsx
		<Component.Root>
			<Component.SubComponent />
			<Component.LowLevelSubComponent lowLevelProp="foo" />
		</Component.Root>
```

Reply:
I like your phrasing, @mirka! Let's bring that up to https://github.com/WordPress/gutenberg/issues/63242 - providing both options sounds like a good compromise that might solve much of the arguments there.

Reply:
I personally thought the monolithic version would be represented by one component, without the need to use compound components.

We'd have:
- the compound component version (as granular or high-level as we'd like)
- the monolithic version (only if we see value in it, or if we have back-compat reasons)

I'm sure about how much value could a "relatively monolithic" approach give. Having an in-between state could cause confusion in the consumers, make naming conventions and APIs more complex.

What value do we see in this approach?

Reply:
I don't really see it as an in-between state, because I don't think most consumers care about a monolithic/modular distinction as they're using a component. They're going to consult the docs anyway, instead of inferring its APIs by whether it's monolithic or modular. It's either just the "default API" that most consumers should use, or the "low-level API" that may be available if further customization is needed.

What matters is that the default API strikes the best balance between consistency, customizability, and convenience, while the low-level API(s) is focused on just customizability. In many cases, we may never need to provide a low-level API at all, but that option is there if we ever do.

And even if we do provide lower-level APIs, we won't necessarily always need the `Component.Root`. For example, if we designed the RadioControl today from scratch, we would definitely pass `RadioControl.Option`s to `children` instead of an object array to the `options` prop. However, if we wanted to also provide a lower-level API we could just expose e.g. a `RadioControl.Radio` and `RadioControl.Label`, without adding a `RadioControl.Root`.

A `Component.Root` only becomes necessary when `<Component>` as a root already includes some functionality that we want to break off as a separate subcomponent.

Reply:
My concern is that offering, at the same time, both a low-level and a high-level version of a compound component could be tricky. We'd have many small components and a much larger API surface to maintain, and we'd need to make sure it's clear which subcomponents can be used together, and which ones don't pair well.

**Have we considered the simpler alternative of dropping the `Root` naming altogether and not committing to offering a high-level version of a compound component?**

- Our compound components would be as low-level as we think is needed
- Less disruption in the naming scheme (no `Root`, less changes with respect to the current situation)
- We can offer more "high-level", ready-made examples in Storybook
- Consumers of the package can create their "high-level" version of the component in their codebases / packages, picking the abstraction they'd prefer
- If a component is already offering a monolithic version of a component, we simply create a new version of the component (`V[n]`) and make it compound.

This would reduce the API surface to maintain, simplify conventions, and it would overall give us a smaller maintenance burden.

Reply:
I agree with your points, and I think we can converge on that.

In general we will set out to design an API that is reasonably low-level, and hopefully will never require adding a `.Root`, as best as we can anticipate. But, if the unanticipated ever happens, we will have `.Root` available as a last resort. Having a separate low-level API is not something we should plan on from the get go.

Does that sound about right? In that case, we don't need to document anything about the `.Root` thing, we can just keep it in our back pocket and pray that we'll never need it.

Reply:
I can write an update in the main issue with this alternative approach, which should hopefully be a better compromise for everyone involved.

Reply:
https://github.com/WordPress/gutenberg/issues/63242#issuecomment-2248422053